### PR TITLE
 :book: proposal for new plugin to generate code ( `deploy-image.go.kubebuilder.io/v1beta1` )

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description**

- EP/proposal to add a new plugin to generate the code requires to address the common needs

**Motivation**

- Many operators will at the end have CRD's which will represent a deployment/pod of an image:tag. So, I am as oper user I'd like to have a feature to make this process easier. Also, see that it address common requests such as https://github.com/operator-framework/operator-sdk/pull/2158
- Reduce the learning curve since users are able to generate the basic operator following the good practices via this feature which makes more intuitive understand how the things work and know-how to improve/develop their solutions.  
- Address common questions over the basic features and how to for example; implement the tests, use envvars, RBAC permissions as watch features. 
- https://github.com/kubernetes-sigs/kubebuilder/issues/2010

### CODE REVIEW

**Comment by pwittrock** [comment]
IIUC -- the motivation is to scaffold projects to operators, which typically run an instance of a container image.  This seems fine to me.  If it can be easily done as a plugin and we will want to support a variety of these sorts of project blueprints, then doing it as a plugin may be best.

Should we support StatefulSet and Deployments?

**Comment by camilamacedo86** [comment]
Hi @pwittrock, 

> Should we support StatefulSet and Deployments?

I thought we started it by using a Deployment. However, we can improve the feature in follow-ups to allow the scaffold use statefulset or webhooks via a flag value such as `kubebuilder create code --image=myexample:0.0.1 --group=example --kind=App --version=v1 --type=[deployment|statefulset|webhook]`. WDYT?

**Comment by pwittrock** [comment]
> I thought we start it by using a Deployment. However, we can improve the feature in follow-ups to support more default types of scaffolds which could be like kubebuilder create code --image=myexample:0.0.1 --group=example --kind=App --version=v1 --type=[deployment|statefulset|webhook]. WDYT?

I agree

**Comment by pwittrock** [comment]
I have no objections to this proposal.

**Comment by camilamacedo86** [comment]
/hold 

Until getting 2 approvals 👍 
For now, we have an LGTM from @pwittrock

**Comment by estroz** [comment]
This feature looks a lot like the existing [`addon`](https://github.com/kubernetes-sigs/kubebuilder/tree/master/plugins/addon) plugin, which eventually should become a phase 2 plugin. Instead of having a separate subcommand it may be worth implementing this in the `create api` subcommand, given the similarities to `addon` and its scaffold dependencies.
```sh
kubebuilder create api --pattern=deployment-image ...
```

**Comment by Adirio** [comment]
I was thinking the same as @estroz, this would only require to scaffold some content inside the `Reconcile` function of the controller and that could be made with a `CreateAPI` plugin in my opinion.

**Comment by camilamacedo86** [comment]
HI @estroz, 

> This feature looks a lot like the existing addon plugin, which eventually should become a phase 2 plugin. Instead of having a separate subcommand it may be worth implementing this in the create api subcommand, given the similarities to addon and its scaffold dependencies.

IMO we should not add in the CreateAPI subcommand because:

- The generated code is not valid for other types of operators. So, Ansible/Helm in SDK, for example, which extending the CREATE API should not have the option/flags to generate the code. 
- The addon should be ported to be a plugin as well and it cannot be an option for other languages such as Ansible/Helm too. See the issue raised in the past which was requested by @DirectXMan12. : https://github.com/kubernetes-sigs/kubebuilder/issues/1543
- To keep better maintainability by not hurting concepts such as single responsibility and cohesion

**Comment by Adirio** [comment]
We are not saying to implement in `CreateAPI` of `go.kubebuilder.io` plugin (the base plugin). We are saying that it can be implemented in another `CreateAPI` plugin, as with phase 2, multiple plugins can be queued once after the other. FIrst you will have `go.kubebuilder.io` scaffolding the file, then `deployment.my-domain.com` will update that file to include the code you suggest. `Addon` will be implemented the same way, probably calling the plugin `addon.kubebuilder.io`.

**Comment by camilamacedo86** [comment]
HI @Adirio and @estroz, 

> We are not saying to implement in CreateAPI of go.kubebuilder.io plugin (the base plugin). We are saying that it can be implemented in another CreateAPI plugin

IMO it cannot be called `create api` since it is doing such more than that. 

> as with phase 2, multiple plugins can be queued once after the other. First you will have go.kubebuilder.io scaffolding the file, then deployment.my-domain.com will update that file to include the code you suggest. Addon will be implemented the same way, probably calling the plugin addon.kubebuilder.io.

I am totally fine with this new plugin/subcommand to be changed and/or cleanup accordingly with phase 2. I'd expected that phase 2 plugins would cover some cleanup and changes as part of its scope too. But I do not think that we need to wait for phase 2 to get it in place or try to define now how it could be after phase 2 since that still very abstract.

**Comment by Adirio** [comment]
The issue I see with `kubebuilder create code` is that it will be deleting a file scaffolded previously by `kubebuilder create api`, which goes against the philosofy we have had about never deleting/removing/breaking code that the user may have written after he used a command.

**Comment by camilamacedo86** [comment]
HI @Adirio, 

> The issue I see with kubebuilder create code is that it will be deleting a file scaffolded previously by kubebuilder create api, which goes against the philosofy we have had about never deleting/removing/breaking code that the user may have written after he used a command.

It is not the idea of this implementation. Note that another reason for POV we are unable to use the Create api is that the `--resource` and `controller` flags do not apply on this one as well.  The create code will never have these options since the API and controller requires to be scaffold as well.  

At the first moment, we will not allow using the command for pre-existent resources which means that by using this feature we will create the resource and controller always. Then, in the future, we could allow use it for pre-existent ones which could be one via markers to just add the code implementation to the files.

**Comment by camilamacedo86** [comment]
Hi @Adirio and @estroz, 

I pushed changes on the EP based on your feedbacks, see the `open questions` section. Really thank your inputs.

**Comment by camilamacedo86** [comment]
Folks, 

I will remove the hold because it has 2 oks already. 
Also, @estroz and @Adirio see that I added all your concerns and points in the open questions section of this EP. 
So, it shows ok to be reviewed and get merged. 

/hold cancel

**Comment by Adirio** [comment]
I'm actually only seeing one ok (from pwittrock). I think you should wait for further apporval.

**Comment by camilamacedo86** [comment]
/hold

After a chat with @Adirio I think we can push changes in this EP before get it merged.

**Comment by Adirio** [comment]
Basically, we were talking that this could be probably be done as a plugin (lets say `deployment.kubebuilder.io/v1`) that wraps `go.kubebuilder.io` and performs the required modifications afterwards until phase 2 is merged. At that point the plugin would be upgraded to `deployment.kubebuilder.io/v2` which no longer wraps `go.kubebuilder.io` but uses the plugin chaining provided by phase 2 to obtains the same result.

**Comment by camilamacedo86** [comment]
HI @Adirio, @estroz, 

It is updated by following your feedback and suggestions over its details in the implementation as well. 
See that the idea here is to be a plugin which will works as addon / declarative one after we merge the solution described and discussed in https://github.com/kubernetes-sigs/kubebuilder/issues/1941

File: designs/code-generate-image-plugin.md
Comment (Adirio) [inline designs/code-generate-image-plugin.md:261]:
```suggestion
To ensure this implementation a new project example should be generated in the [testdata](../testdata/) directory of the project. See the [test/testadata/generate.sh](../test/testadata/generate.sh). Also, we should use this scaffold in the [integration tests](../test/e2e/) to ensure that the data scaffolded with works on the cluster as expected.  
```

File: designs/code-generate-image-plugin.md
Comment (Adirio) [inline designs/code-generate-image-plugin.md:275]: Needs to be a list because it may be used for several resources:
```suggestion
plugins:
    generate.kubebuilder.io/v1:
        resources:
          - domain: example.io
            group: crew
            kind: Captain
            version: v1
            image: "<some-registry>/<project-name>:<tag>
```

**Review by Adirio — CHANGES_REQUESTED** [review]
Disclaimer: I didn't read thoroughly the implementation examples.

I agree with the goals of this EP. My two cents on the open questions at the end:

1. **Should we allow to scaffold the code for an API that is already created for the project?**
I agree that we should require both the API and the controller to be created with the same command call as both parts are required for this plugin.

2. **Should we support StatefulSet and Deployments?**
Starting with one and increasing the supported types incrementally sounds good to me.

3. **Could this feature be useful to other languages or is it just valid to Go based operators?**
This plugin would mainly modify `*_types.go` and `*_controller.go` which are obviously only relevant for go based operators. In a future, if other languaged based operators are supported (either officially or by the community) this plugin could be used as reference to create an equivalent one for their language. Therefore, it should probably be a "subdomain" of `go.kubebuilder.io`

NIT: I think we need to come up with a better name for this plugin. Generate code seems pretty imprecise in a scaffolding project. The plugin is about good practices and image deploying. Do you like `good-practices.go.kubebuilder.io` or `deploy-image.go.kubebuilder.io`?  I don't really like the first for a plugin name despite being descriptive.

**Comment by camilamacedo86** [comment]
Hi @Adirio, 

Thank you for your comments and review. I adopted the name `deploy-image.go.kubebuilder.io` suggested and addressed your comments as well.

**Review by Adirio — APPROVED** [review]
Will let other mantainers the chance to take a look before merging, but LGTM

File: designs/code-generate-image-plugin.md
Comment (estroz) [inline designs/code-generate-image-plugin.md:289]: Since kubebuilder is only concerned with Go projects and the proposal discusses Go-only changes, I think the answer to this question is simply "no".

File: designs/code-generate-image-plugin.md
Comment (estroz) [inline designs/code-generate-image-plugin.md:270]: This configuration implies that phase 2 is required to implement this plugin, since the 1.5 proposal doesn't discuss adding plugin configs to `plugins`. @Adirio that's correct right?

**Review by estroz — CHANGES_REQUESTED** [review]
Overall lgtm, and I agree with the responses @Adirio has in https://github.com/kubernetes-sigs/kubebuilder/pull/1803#pullrequestreview-590251624.

A few things before I approve this proposal:
1. This proposal is dependent on at least phase 1.5, if not phase 2, plugins, so it can't be implemented or tested until the that requirement is complete.
2. Will this plugin live in the kubebuilder repo? I am ok with that, but the test matrix will start blowing up if more plugins are added. Something to be aware of. IMO I'd like to see a `kubebuilder-plugins` repo, with submodules for each plugin.

File: designs/code-generate-image-plugin.md
Comment (Adirio) [reply designs/code-generate-image-plugin.md:270]: project-version v3 already supports that field. Nothing specific was defined for plugin phase 1.5 but it doesn't prevent it either. So I would say it doesn't require phase 2 (supporting external binaries is something we still haven't properly designed AFAIK) but it definetely requires phase 1.5.

**Comment by Adirio** [comment]
> 2. Will this plugin live in the kubebuilder repo? I am ok with that, but the test matrix will start blowing up if more plugins are added. Something to be aware of. IMO I'd like to see a `kubebuilder-plugins` repo, with submodules for each plugin.

Camila and me have already recognized the test matrix issue (there are already some permutations that we are not testing like plugin go/v2 with project-version v3) and adding plugins will make it even a greater issue. I already considered a separate repo (I called it `kubebuilder-community` but `kubebuilder-plugins` may be a better name.

File: designs/code-generate-image-plugin.md
Comment (camilamacedo86) [reply designs/code-generate-image-plugin.md:289]: I would like to leave the answer to clarify some questions raised. This plugin is only valid for go. However, the same idea could be used as an example.

**Comment by camilamacedo86** [comment]
Hi @Adirio, @estroz, 

> Camila and me have already recognized the test matrix issue (there are already some permutations that we are not testing like plugin go/v2 with project-version v3) and adding plugins will make it even a greater issue. I already considered a separate repo (I called it kubebuilder-community but kubebuilder-plugins may be a better name)

> Will this plugin live in the kubebuilder repo? I am ok with that, but the test matrix will start blowing up if more plugins are added. Something to be aware of. IMO I'd like to see a kubebuilder-plugins repo, with submodules for each plugin.

This plugin address a need that has been requested in SDK for mone than one year. For example https://github.com/operator-framework/operator-sdk/pull/2158. Also, it addresses some discussions made with @pwittrock as well in order to make the things easier to end-users address the most common causes. So, this one would be born with "official/beta" support.   

And then, if you check the https://github.com/kubernetes-sigs/kubebuilder/issues/2016 the plugin phase 2.0 will look would be responsible to only check the possibility of the plugins be as separate binaries that can be discovered by the Kubebuilder CLI binary via user-specified plugin file paths. 

In this way, now we have not too many plugins. So, shows Premature Optimization care about. Let's think about that after we have at least 5 plugins and after the phase 2. Is it make sense? can we agree on that?

**Review by estroz — APPROVED** [review]
Despite the open issues, this is ok for now.

/lgtm

**Pull request merged**
